src/formats/CML.cpp

```diff
@@ -0,0 +1,581 @@
+// Chemfiles, a modern library for chemistry file reading and writing
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#include <cassert>
+#include <sstream>
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include "pugixml.cpp"
+
+#include "chemfiles/formats/CML.hpp"
+
+#include "chemfiles/ErrorFmt.hpp"
+#include "chemfiles/File.hpp"
+#include "chemfiles/Frame.hpp"
+#include "chemfiles/utils.hpp"
+#include "chemfiles/warnings.hpp"
+#include "chemfiles/periodic_table.hpp"
+
+using namespace chemfiles;
+
+template<> FormatInfo chemfiles::format_information<CMLFormat>() {
+    return FormatInfo("CML").with_extension(".cml").description(
+        "Chemical Markup Language"
+    );
+}
+
+CMLFormat::CMLFormat(std::string path, File::Mode mode, File::Compression compression)
+    : mode_(mode), file_(TextFile::open(std::move(path), mode, compression))
+{
+
+    if (mode_ == File::WRITE) {
+        root_ = document_.append_child("cml");
+        root_.append_attribute("xmlns") = "http://www.xml-cml.org/schema";
+        root_.append_attribute("xmlns:cml") = "http://www.xml-cml.org/dict/cml";
+        root_.append_attribute("xmlns:units") = "http://www.xml-cml.org/units/units";
+        root_.append_attribute("xmlns:convention") = "http://www.xml-cml.org/convention";
+        root_.append_attribute("convention") = "convention:molecular";
+        root_.append_attribute("xmlns:iupac") = "http://www.iupac.org";
+        return;
+    }
+
+    if (mode_ == File::APPEND) { // may need to check the file a CML node in the future, now just append
+        root_ = document_;
+        return;
+    }
+
+    auto result = document_.load(*file_);
+    if (!result) {
+        throw format_error("[CML] Parsing error: '{}'", result.description());
+    }
+
+    root_ = document_.child("cml");
+    if (root_) {
+        auto molecules = root_.children("molecule");
+        current_ = molecules.begin();
+        if (current_ == molecules.end()) {
+            throw format_error("[CML] cml node has no valid children");
+        }
+        return;
+    }
+
+    // Not technically standard to have multiple molecule nodes as root nodes,
+    // but some tools do this and we should support it 
+    if (document_.child("molecule")) {
+        auto molecules = document_.children("molecule");
+        current_ = molecules.begin();
+        root_ = document_;
+        return;
+    }
+
+    throw format_error("[CML] no supported starting node found");
+}
+
+CMLFormat::~CMLFormat() {
+    if (num_added_ && mode_ == File::WRITE) {
```

Comment:
`if (num_added_ != 0 && mode_ == File::WRITE) {` is recommanded by clang-tidy. You can run it by building the code with clang and setting `CHFL_USE_CLANG_TIDY=ON` with cmake